File: src/utils/logger.ts
export class Logger {
  private prefix: string
  private logLevel: 'debug' | 'info' | 'warn' | 'error'

  constructor(
    prefix: string,
    logLevel: 'debug' | 'info' | 'warn' | 'error' = 'info'
  ) {
    this.prefix = prefix
    this.logLevel = import.meta.env.PROD ? 'info' : logLevel
  }

  debug(message: string, ...args: any[]): void {
    if (this.logLevel === 'debug')
      console.debug(`[${this.prefix}] ${message}`, ...args)
  }

  info(message: string, ...args: any[]): void {
    if (this.logLevel === 'debug' || this.logLevel === 'info')
      console.log(`[${this.prefix}] ${message}`, ...args)
  }

  warn(message: string, ...args: any[]): void {
    if (this.logLevel !== 'error')
      console.warn(`[${this.prefix}] ${message}`, ...args)
  }

  error(message: string, ...args: any[]): void {
    console.error(`[${this.prefix}] ${message}`, ...args)
  }
}

File: python/utils/logger.py
"""
Logging configuration for Alice AI Backend
"""

import logging
import sys
from pathlib import Path
from typing import Optional

from config import settings


class ColoredFormatter(logging.Formatter):
    """Colored log formatter for better console output."""
    
    # ANSI color codes
    COLORS = {
        'DEBUG': '\033[36m',     # Cyan
        'INFO': '\033[32m',      # Green
        'WARNING': '\033[33m',   # Yellow
        'ERROR': '\033[31m',     # Red
        'CRITICAL': '\033[35m',  # Magenta
    }
    RESET = '\033[0m'
    
    def format(self, record):
        # Add color to levelname
        if record.levelname in self.COLORS:
            record.levelname = f"{self.COLORS[record.levelname]}{record.levelname}{self.RESET}"
        
        # Format the message
        return super().format(record)


def setup_logger(name: str, log_file: Optional[str] = None) -> logging.Logger:
    """Set up a logger with consistent formatting."""
    
    logger = logging.getLogger(name)
    
    # Don't add handlers if already configured
    if logger.handlers:
        return logger
    
    logger.setLevel(getattr(logging, settings.log_level.upper()))
    
    # Create console handler with colored output and proper encoding
    # Force UTF-8 encoding on Windows to handle Unicode characters
    if sys.platform.startswith('win'):
        import io
        console_handler = logging.StreamHandler(
            io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')
        )
    else:
        console_handler = logging.StreamHandler(sys.stdout)
    console_handler.setLevel(logging.DEBUG)
    
    # Use colored formatter for console
    console_formatter = ColoredFormatter(
        fmt='%(asctime)s [%(levelname)s] %(name)s: %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )
    console_handler.setFormatter(console_formatter)
    logger.addHandler(console_handler)
    
    # Add file handler if specified
    if log_file:
        file_handler = logging.FileHandler(log_file)
        file_handler.setLevel(logging.DEBUG)
        
        # Use plain formatter for file
        file_formatter = logging.Formatter(
            fmt='%(asctime)s [%(levelname)s] %(name)s: %(message)s',
            datefmt='%Y-%m-%d %H:%M:%S'
        )
        file_handler.setFormatter(file_formatter)
        logger.addHandler(file_handler)
    
    return logger


def get_logger(name: str) -> logging.Logger:
    """Get a logger instance."""
    return logging.getLogger(name)